multicluster/controllers/multicluster/core/resourceimport_controller.go

```diff
@@ -160,25 +174,27 @@ func (r *ResourceImportReconciler) handleResImpUpdateForService(ctx context.Cont
 		return ctrl.Result{}, err
 	}
 	svcImpObj := getMCServiceImport(resImp)
+	if !reflect.DeepEqual(*svc, corev1.Service{}) {
+		svcImpObj.Spec.IPs = []string{svc.Spec.ClusterIP}
+	}
 	if apierrors.IsNotFound(err) {
-		creationSucceed = false
 		err := r.localClusterManager.Create(ctx, svcImpObj, &client.CreateOptions{})
 		if err != nil {
 			klog.ErrorS(err, "fail to create ServiceImport", "serviceimport", klog.KObj(svcImpObj))
 			return ctrl.Result{}, err
 		}
-		creationSucceed = true
+		svcImpCreated = true
 	}
 
-	if creationSucceed {
+	if svcCreated && svcImpCreated {
 		r.installedResImports.Add(*resImp)
+		if len(svcImpObj.Spec.IPs) == 0 {
+			// requeue the event to update ServiceImport's ClusterSetIP
+			return ctrl.Result{}, fmt.Errorf("ServiceImport %s ClusterSetIP is empty", klog.KObj(svcImpObj))
```

Comment:
Humm - how long it takes for the retry? Thinking whether we should optimize by watching Services to get the IP faster but I know it is much complexer.

Reply:
I checked the codes, it seems using the default rate limit here https://github.com/kubernetes/client-go/blob/master/util/workqueue/default_rate_limiters.go#L39, interval should be from 5ms to 10s. if multiple failures happened, the interval will be like `baseDelay*2^<num-failures>`, `baseDelay` is 5ms, not sure if this kind of interval is enough or not, I will check if I can add more logic in service_controller to watch Service change, then update the IP

Reply:
I am still trying to understand if we need a Service controller or not. If not, could we think about a simpler solution to handle this, like some way to requeue the event after a short interval?

Reply:
service controller only watch those service with mcs annotation, so I suppose it should be OK to have one? I haven't found a valid way to handle the case when ResourceImport is deleted but event is not handled by controller in member cluster. I will check if reconciler can run like RunOnce().

Reply:
@jianjuns  about requeue for ClusterSetIP setting, when we return empty result with error, the event will actually be requeued,  do you mean the interval controlled by default rate limit is not short enough?

Reply:
If it is always ms to seconds, that should be good enough.

Reply:
@jianjuns I removed service_controller and add a standard controller named stale_controller which will run only once if no error happens during stale Service/ServiceImport clean up. let me know if you have more comments. thanks.

Reply:
I think better to get @tnqn 's suggestion here. Could you check with him?

Reply:
sure, I have pinged him directly, let's see if there is comment from @tnqn